# Tarefa 02 Módulo 05

O nosso projeto desta sequência de módulos do curso será um aprofundamento da demonstração sobre classificação de risco de crédito que vimos lá no comecinho. Pois recebemos uma base já montada pra nós. Tenha certeza de que ela passou por um longo processamento até ficar daquele jeito. Neste exercício vamos exercitar o que aprendemos nas ultimas aulas e montar a variável resposta da base do nosso projeto.

#### Marcação de bom e mau
O objetivo da modelagem é classificar o risco de inadimplência, ou como se diz no meio, o risco de *default*. Podemos fazer longas discussões sobre o conceito de *default* com base em estudos e exigências regulatórias, para efeitos deste estudo, um cliente em *default* é aquele que está em 60 dias de atraso ou mais. Então classificaremos os clientes como 'bons' e 'maus' assim:
- **Maus** pagadores: são aqueles que entraram em 'default' (atraso 60 dias ou mais) nos 24 meses seguintes à aquisição do cartão de crédito. 
- **Bons** pagadores: são considerados todos os demais.
- **Excluídos**: Clientes que não adquiriram um cartão de crédito (seja por recusa, seja por desistência) não possuem informações de pagamento, portanto não se pode identificar se são bons ou maus. Há uma longa discussão e literatura sobre *inferência de rejeitados* que está fora do escopo deste exercício.

#### Bases disponíveis
Temos duas bases importantes aqui: uma de propostas, com diversas informações dos vários solicitantes de cartão de crédito, e uma base de pagamentos. A base de pagamentos será utilizada para identificar a ocorrência de *default*. A base de propostas tem diversas informações coletadas no momento da solicitação do crédito (isto é importante: qualquer informação posterior a essa data é impossível de ser coletada na aplicação do modelo e não pode ser utilizada).

As variáveis delas são:

Base de propostas - application_records.csv

| Nome da Variável         | Description                                         | Tipo  |
| ------------------------ |:---------------------------------------------------:| -----:|
| ID| identificador do cliente (chave) |inteiro|
| CODE_GENDER| M = 'Masculino'; F = 'Feminino' |M/F|
| FLAG_OWN_CAR| Y = 'possui'; N = 'não possui' |Y/N|
| FLAG_OWN_REALTY| Y = 'possui'; N = 'não possui' |Y/N|
| CNT_CHILDREN| Quantidade de filhos |inteiro|
| AMT_INCOME_TOTAL| Annual income |inteiro|
| NAME_INCOME_TYPE|Tipo de renda (ex: assaliariado, autônomo etc) | texto |
| NAME_EDUCATION_TYPE| Nível de educação (ex: secundário, superior etc) |texto|
| NAME_FAMILY_STATUS | Estado civil (ex: solteiro, casado etc)| texto |
| NAME_HOUSING_TYPE | tipo de residência (ex: casa/apartamento, com os pais etc) | texto |
| DAYS_BIRTH | Count backwards from current day (0), -1 means yesterday |inteiro|
| DAYS_EMPLOYED | Count backwards from current day (0), -1 means yesterday |inteiro|
| FLAG_MOBIL | Indica se possui celular (1 = sim, 0 = não) |binária|
| FLAG_WORK_PHONE | Indica se possui telefone comercial (1 = sim, 0 = não) |binária|
| FLAG_PHONE | Indica se possui telefone (1 = sim, 0 = não) |binária|
| FLAG_EMAIL | Indica se possui e-mail (1 = sim, 0 = não) |binária|
| OCCUPATION_TYPE | Occupation	 |Qualitativa|
| CNT_FAM_MEMBERS | quantidade de pessoas na residência |inteiro|

Base de pagamentos - pagamentos_largo.csv  

| Nome da Variável         | Description                                         | Tipo  |
| ------------------------ |:---------------------------------------------------:| -----:|
| ID| identificador do cliente (chave) |inteiro|
| mes_00 a mes_24| faixa de atraso mês a mês do cliente <br>0: 1-29 days past due &nbsp;&nbsp;&nbsp;&nbsp; 1: 30-59 days past due <br />2: 60-89 days overdue &nbsp;&nbsp;&nbsp;&nbsp; 3: 90-119 days overdue <br /> 4: 120-149 days overdue &nbsp;&nbsp;&nbsp;&nbsp; 5: more than 150 days <br />C: paid off that month &nbsp;&nbsp;&nbsp;&nbsp; X: No loan for the month |Qualitativa|

#### Construindo a variável resposta
A base de pagamentos está em um formato de 'base larga'. Essa base possui informações de pagamentos do cliente mês a mês a partir do mês de aquisição do crédito (mês 0) até o vigésimo quarto mês após a aquisição do crédito (mês 24). Utilizaremos essa base para determinar se um proponente é considerado 'bom pagador' ou caso apresente atraso representativo, será considerado 'mau pagador'.

#### Base larga vs base longa
A base ser larga significa que há uma linha para cada cliente, e que as informações estarão nas colunas, em contraste com a 'base longa', em que haveria uma linha para cada combinação cliente/mês, uma coluna indicando o cliente, outra indicando o mês, e apenas uma coluna com a informação do atraso.

#### Tarefa 1) Marcar *default* no mês
Faça uma indicadora de se o cliente está em *default* em cada uma das marcações (mes_00 a mes_24). Dica: você pode utilizar o método ```.isin()``` do Pandas. Consulte a [documentação](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isin.html) caso necessário.

#### Tarefa 2) 'bons' e 'maus' ao longo de todos os 24 meses de desempenho
Marque para cada cliente se ele teve pelo menos um episódio de *default* entre o mês 0 e o mês 24. Dica: o método ```sum()``` pode ajudar. Caso precise, consulte a [documentação](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sum.html) e procure pelo argumento ```axis```, você viu outros métodos que possuem esse argumento também. Tendo o número de meses em default de cada cliente, basta marcar ```True``` para todos aqueles que possuem pelo menos 1 mês em *default* e ```False``` para os demais.

#### Tarefa 3) Marcando proponentes expostos ao risco de crédito
Marcando proponentes que se tornaram tomadores: lembre-se de que clientes que não adquiriram o cartão devem ser desconsiderados. A base de pagamentos possui apenas clientes que adquiriram cartão de crédito, então você pode selecionar somente os clientes da base de propostas que se encontram na base de pagamentos.

#### Tarefa 4) Consolidando as informações
Faça uma junção das informações da base de propostas com a variável de *default* que você acabou de construir. Talvez você consiga realizar a tarefa 3 e tarefa 4 em uma única linha de código ;)

#### Tarefa 5) Verificando
Faça uma contagem dos valores do *default* que você construiu. 

In [2]:
import pandas as pd

In [271]:
propostas = pd.read_csv('application_record.csv')
pg = pd.read_csv('pagamentos_largo.csv')

In [277]:
# Organizar o DataFrame para ter os meses em ordem

pg1 = pg[['ID', 'mes_0', 'mes_1', 'mes_2', 'mes_3', 'mes_4', 'mes_5', 'mes_6', 'mes_7', 'mes_8', 'mes_9', 'mes_10', 'mes_11',
            'mes_12', 'mes_13', 'mes_14', 'mes_15', 'mes_16', 'mes_17', 'mes_18', 'mes_19', 'mes_20', 'mes_21', 'mes_22', 'mes_23', 'mes_24']]

pg1 = pg1.dropna()
i = 0
default_pg = pd.DataFrame()
default_pg['ID'] = pg1['ID']
while i < 25:
    default_pg[f'default_mes_{i}'] = (pg1[f'mes_{i}'].isin(['2', '3', '4', '5'])).astype('int8')
    i += 1


In [278]:
# 1) Seu código aqui

default_pg

ID  default_mes_0  default_mes_1  default_mes_2  default_mes_3  \
1      5001719              0              0              0              0   
2      5001720              0              0              0              0   
4      5001726              0              0              0              0   
5      5001730              0              0              0              0   
6      5001738              0              0              0              0   
...        ...            ...            ...            ...            ...   
20929  5150459              0              0              0              0   
20931  5150473              0              0              0              0   
20932  5150475              0              0              0              0   
20934  5150480              0              0              0              0   
20936  5150487              0              0              0              0   

       default_mes_4  default_mes_5  default_mes_6  default_mes_7  \
1                  0              0              0              0   
2                  0              0              0              0   
4                  0              0              0              0   
5                  0              0              0              0   
6                  0              0              0              0   
...              ...            ...            ...            ...   
20929              0              0              0              0   
20931              0              0              0              0   
20932              0              0              0              0   
20934              0              0              0              0   
20936              0              0              0              0   

       default_mes_8  ...  default_mes_15  default_mes_16  default_mes_17  \
1                  0  ...               0               0               0   
2                  0  ...               0               0               0   
4                  0  ...               0               0               0   
5                  0  ...               0               0               0   
6                  0  ...               0               0               0   
...              ...  ...             ...             ...             ...   
20929              0  ...               0               0               0   
20931              0  ...               0               0               0   
20932              0  ...               0               0               0   
20934              0  ...               0               0               0   
20936              0  ...               0               0               0   

       default_mes_18  default_mes_19  default_mes_20  default_mes_21  \
1                   0               0               0               0   
2                   0               0               0               0   
4                   0               0               0               0   
5                   0               0               0               0   
6                   0               0               0               0   
...               ...             ...             ...             ...   
20929               0               0               0               0   
20931               0               0               0               0   
20932               0               0               0               0   
20934               0               0               0               0   
20936               0               0               0               0   

       default_mes_22  default_mes_23  default_mes_24  
1                   0               0               0  
2                   0               0               0  
4                   0               0               0  
5                   0               0               0  
6                   0               0               0  
...               ...             ...             ...  
20929               0               0               

In [279]:
# 2) Seu código aqui

default_pg.loc[default_pg.drop(columns=['ID']).sum(axis=1) == 0, 'bom_mau'] = 'bom'
default_pg['bom_mau'].fillna('mau', inplace=True)
pd.get_dummies(default_pg["bom_mau"]).astype("int8")

bom  mau
1        1    0
2        1    0
4        1    0
5        1    0
6        1    0
...    ...  ...
20929    1    0
20931    1    0
20932    1    0
20934    1    0
20936    1    0

[12267 rows x 2 columns]

In [280]:
# 3) Seu código aqui

propostas[propostas['FLAG_OWN_CAR'] == "Y"]

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008804           M            Y               Y             0   
1       5008805           M            Y               Y             0   
2       5008806           M            Y               Y             0   
10      5008815           M            Y               Y             0   
11      5112956           M            Y               Y             0   
...         ...         ...          ...             ...           ...   
438519  6836039           M            Y               N             0   
438539  6837454           M            Y               Y             1   
438542  6837905           M            Y               Y             1   
438543  6837906           M            Y               Y             1   
438548  6839936           M            Y               Y             1   

        AMT_INCOME_TOTAL NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
0               427500.0          Working               Higher education   
1               427500.0          Working               Higher education   
2               112500.0          Working  Secondary / secondary special   
10              270000.0          Working               Higher education   
11              270000.0          Working               Higher education   
...                  ...              ...                            ...   
438519          157500.0          Working               Higher education   
438539          162000.0    State servant  Secondary / secondary special   
438542          355050.0          Working  Secondary / secondary special   
438543          355050.0          Working  Secondary / secondary special   
438548          135000.0          Working  Secondary / secondary special   

       NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
0          Civil marriage   Rented apartment      -12005          -4542   
1          Civil marriage   Rented apartment      -12005          -4542   
2                 Married  House / apartment      -21474          -1134   
10                Married  House / apartment      -16872           -769   
11                Married  House / apartment      -16872           -769   
...                   ...                ...         ...            ...   
438519            Married  House / apartment      -11619          -4149   
438539            Married  House / apartment      -10890          -2675   
438542            Married  House / apartment      -15904          -2614   
438543            Married  House / apartment      -15904          -2614   
438548            Married  House / apartment      -12569          -2095   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
0                1                1           0           0             NaN   
1                1                1           0           0             NaN   
2                1                0           0           0  Security staff   
10               1                1           1           1     Accountants   
11               1                1           1           1     Accountants   
...            ...              ...         ...         ...             ...   
438519           1                0           0           0      Core staff   
438539           1                0           0           0      Core staff   
438542           1                0           0           0             NaN   
438543           1                0           0           0             NaN   
438548           1                0           0           0        Laborers   

        CNT_FAM_MEMBERS  
0                   2.0  
1                   2.0  
2                   2.0  
10                  2.0  
11                  2.0  
...                 ...  
438519              2.0  
438539              3.0  
438542              3.0  
438543              3.0  
438548              3.0  

[163098 rows x 18 columns]

In [281]:
# 4) eu código aqui

com_cartao = propostas.merge(default_pg, on='ID', how ='left')
com_cartao = com_cartao[com_cartao['FLAG_OWN_CAR'] == "Y"]
com_cartao['OCCUPATION_TYPE'].fillna('NOJOB', inplace=True)
com_cartao = com_cartao.dropna()
com_cartao

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
19      5008825           F            Y               N             0   
20      5008826           F            Y               N             0   
35      5008844           M            Y               Y             0   
37      5008847           M            Y               Y             0   
39      5008850           M            Y               Y             0   
...         ...         ...          ...             ...           ...   
434747  5118443           F            Y               N             1   
434773  5139311           M            Y               Y             0   
434775  5142185           M            Y               Y             2   
434803  5149145           M            Y               Y             0   
434804  5149158           M            Y               Y             0   

        AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
19              130500.0               Working              Incomplete higher   
20              130500.0               Working              Incomplete higher   
35              112500.0  Commercial associate  Secondary / secondary special   
37              112500.0  Commercial associate  Secondary / secondary special   
39              112500.0  Commercial associate  Secondary / secondary special   
...                  ...                   ...                            ...   
434747          108000.0  Commercial associate               Higher education   
434773          180000.0         State servant               Higher education   
434775          450000.0  Commercial associate               Higher education   
434803          247500.0               Working  Secondary / secondary special   
434804          247500.0               Working  Secondary / secondary special   

       NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  ...  default_mes_16  \
19                Married  House / apartment  ...             0.0   
20                Married  House / apartment  ...             0.0   
35                Married  House / apartment  ...             0.0   
37                Married  House / apartment  ...             0.0   
39                Married  House / apartment  ...             0.0   
...                   ...                ...  ...             ...   
434747            Married  House / apartment  ...             0.0   
434773            Married  House / apartment  ...             0.0   
434775            Married  House / apartment  ...             0.0   
434803            Married  House / apartment  ...             0.0   
434804            Married  House / apartment  ...             0.0   

        default_mes_17  default_mes_18  default_mes_19  default_mes_20  \
19                 0.0             0.0             0.0             0.0   
20                 0.0             0.0             0.0             0.0   
35                 0.0             0.0             0.0             0.0   
37                 0.0             0.0             0.0             0.0   
39                 0.0             0.0             0.0             0.0   
...                ...             ...             ...             ...   
434747             0.0             0.0             0.0             0.0   
434773             0.0             0.0             0.0             0.0   
434775             0.0             0.0             0.0             0.0   
434803             0.0             0.0             0.0             0.0   
434804             0.0             0.0             0.0             0.0   

        default_mes_21 default_mes_22  default_mes_23  default_mes_24  bom_mau  
19                 0.0            0.0             0.0             0.0      bom  
20                 0.0            0.0             0.0             0.0      bom  
35                 0.0            0.0             0.0             0.0      bom  
37                 0.0            0.0             0.0             0.0      bom  
39                 0.0      

In [283]:
# 5) Verificando
# Faça uma contagem dos valores do default que você construiu.

pd.get_dummies(com_cartao["bom_mau"]).astype("int8").sum()

bom    3567
mau     102
dtype: int64

 Logo da pra ver que os bons clientes, que tem cartão e que estão adimplentes são 3567.
 E os clientes inadimplentes (default), que também tem cartão são 102.